homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,408 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Iterable
+import json
+import logging
+from types import FunctionType
+from typing import Any
+
+import voluptuous as vol
+from voluptuous_openapi import UNSUPPORTED, convert
+
+from homeassistant.core import Context, HomeAssistant, State, callback
+from homeassistant.util import dt as dt_util
+
+from . import (
+    area_registry as ar,
+    config_validation as cv,
+    entity_registry as er,
+    floor_registry as fr,
+    intent,
+    template,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_KEY = "llm_tool"
+
+NO_PARAMETERS = {"type": "object", "properties": {}, "required": []}
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str
+    parameters: dict[str, Any]
```

Comment:
My impression is that tool agents expecting an openapi description of the tool would want to have a description of the response as well: https://cloud.google.com/dialogflow/vertex/docs/concept/tools#openapi

(We could also consider responses as a second pass if that helps make incremental progress.)

Reply:
No, the Python API only uses input parameters: https://ai.google.dev/api/python/google/ai/generativelanguage/FunctionDeclaration. Also LLMs are quite good at understanding structured json tool output in any format.